MDEV-30975 Wrong result with cross Join given join order

DaveGosselin-MariaDB opened the pull request:
For queries with derived tables populated having some side-effect, we will fill such a derived table more than once, but without clearing its rows.  Consequently it will have duplicate rows. An example query exhibiting the problem is
  SELECT STRAIGHT_JOIN c1 FROM t1 JOIN (SELECT @a := 0) x;
Since mysql_derived_fill will, for UNCACHEABLE_DEPENDENT tables, drop all rows and repopulate, we relax the condition within mysql_derived_fill: rather than assume all uncacheable values prevent an early return, we now allow an early return for uncacheable values other than UNCACHEABLE_DEPENDENT.  In general, we only populate derived tables once unless they're dependent tables.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=2999) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=2999) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=2999) it.</sub>

cvicentiu commented on sql/sql_derived.cc:
```diff
@@ -1259,6 +1261,7 @@ bool mysql_derived_fill(THD *thd, LEX *lex, TABLE_LIST *derived)
   }
   else
   {
+    assert (!unit->executed || (unit->uncacheable & UNCACHEABLE_DEPENDENT));
```
use DBUG_ASSERT()

spetrunia reviewed (approved):
Please act on @cvicentiu's input
Then apply this change to fix coding style and make the testcase contain a clue about what is being testsed:

https://gist.github.com/spetrunia/9d0cffde72308ea904c2f372992b6e08

Ok to push after that is addressed.

DaveGosselin-MariaDB commented:
I created a new PR to merge to 10.5 and I'm closing this PR as a result.  The new PR is https://github.com/MariaDB/server/pull/3062 which I will merge once the buildbot checks complete as it addresses the comments made here.
